### PR TITLE
refactor: Imporve signer confirm tx

### PR DESCRIPTION
**Pull request by htiennv**
This PR using a `Ticker` for periodically call `GetTx` by given hash instead of a `Timer`. This makes the code clean and avoids calling reset `Timer` multiple times.

It's similar to `WaitMined` in [geth](https://github.com/ethereum/go-ethereum/blob/09e0208029ff96a9cda0c69dbaebfd3f31a39771/accounts/abi/bind/util.go#L32).

Other changes:
* Add lock when get `pollTime`

### CODE REVIEW

File: pkg/user/signer.go [hunk 1]
```diff
@@ -257,8 +253,6 @@ func (s *Signer) Address() sdktypes.AccAddress {
 
 // SetPollTime sets how often the signer should poll for the confirmation of the transaction
 func (s *Signer) SetPollTime(pollTime time.Duration) {
-	s.mtx.Lock()
-	defer s.mtx.Unlock()
 	s.pollTime = pollTime
 }
```
[the remark this answers is not in the document]
Comment (htiennv) [reply pkg/user/signer.go]: Done

**Review by evan-forbes — DISMISSED** [review]
lgtm

File: pkg/user/signer.go [hunk 2]
```diff
@@ -153,7 +153,7 @@ func (s *Signer) SubmitPayForBlob(ctx context.Context, blobs []*blob.Blob, opts
 		return nil, err
 	}
 	if resp.Code != 0 {
-		return resp, fmt.Errorf("tx failed with code %d: %s", resp.Code, resp.RawLog)
+		return nil, fmt.Errorf("tx failed with code %d: %s", resp.Code, resp.RawLog)
```
Comment (evan-forbes) [inline pkg/user/signer.go:156]: blocking

wait this change seems unrelated. why do we want to return nil and not the response here?

Comment (htiennv) [reply pkg/user/signer.go:156]: Thanks @evan-forbes for your review!
I see here we are already returning an error with a code, so there is no need to return a response. Returning nil would be better.

Comment (htiennv) [reply pkg/user/signer.go:156]: And I don't understand why the old code returns response here.

Comment (evan-forbes) [reply pkg/user/signer.go:156]: your point is definitley valid, where ideally we would return only the tx response or nil and an error. Since the sdk includes so much information in the tx response, I think we need to keep returning it as well. We can't get rid of the error entirely either since there are instances where the sdk will return a nil tx response. Perhaps we can not perform the check on the code to keep things logically cleaner, however that check is performed in almost all cases.

regardless, that change is outside of the scope of the refactor imo, and changing the return value seems very close to a breaking change if it isn't one. Existing scripts that I've written would at least result in a panic should this get hit.

Comment (evan-forbes) [reply pkg/user/signer.go:156]: we can see this actually break the `TestLegacyUpgrade` test

Comment (evan-forbes) [reply pkg/user/signer.go:156]: I'd approve without this change, and we should open an issue to track the returning nil thing

Comment (htiennv) [reply pkg/user/signer.go:156]: Ok, thanks! I will ignore this change.

Comment (htiennv) [reply pkg/user/signer.go]: Done @evan-forbes

File: pkg/user/signer.go [hunk 3]
```diff
@@ -153,6 +153,7 @@ func (s *Signer) SubmitPayForBlob(ctx context.Context, blobs []*blob.Blob, opts
 		return nil, err
 	}
 	if resp.Code != 0 {
+		// TODO(htienv): Should return nil here?
```
Comment (rootulp) [inline pkg/user/signer.go:156]: good question. It seems odd to return a response and an error from this function.

Comment (htiennv) [reply pkg/user/signer.go:156]: Yes, I think need reconsider for ensure it doesn't affect other things

Comment (rootulp) [reply pkg/user/signer.go:156]: I'm not strongly opinionated on this but do you mind removing your handle from the TODO comment?

```suggestion
		// TODO: should this return nil? 
```

or removing the line all together and creating a GH issue?

**Review by rootulp — COMMENTED** [review]
[question] IDK how this was missed in a previous PR but should https://github.com/htiennv/celestia-app/blob/563a154b4f76e1489bda929b64024677117723ff/pkg/user/signer_test.go#L72 contain a leading `Test` so that it actually runs in the test suite?

**Review by evan-forbes — COMMENTED** [review]
it looks like we're still seeing quite a few tests fail, so we need to figure out why before we can merge anything. per the other review we should also clean up todo comments

File: pkg/user/signer.go (hunk 3 above)
Comment (htiennv) [reply pkg/user/signer.go:156]: Yes, thank you!

**Comment by htiennv** [comment]
Oh sorry, I will fix this PR for handle test cases!

**Comment by htiennv** [comment]
I have update this PR, please review it @evan-forbes @rootulp. Thanks!

File: pkg/user/signer.go [hunk 4]
```diff
@@ -206,36 +206,35 @@ func (s *Signer) BroadcastTx(ctx context.Context, txBytes []byte) (*sdktypes.TxR
...
-				}
-				return resp.TxResponse, nil
-			}
-
-			if !strings.Contains(err.Error(), "not found") {
-				return &sdktypes.TxResponse{}, err
-			}
-
-			timer.Reset(s.pollTime)
+		case <-poolTicker.C:
```
[the remark this answers is not in the document]
Comment (htiennv) [reply pkg/user/signer.go]: Thanks

**Comment by rootulp** [comment]
Thanks for the ping @htiennv . I'm observing that `TestConfirmTxTimeout` fails so I pushed a commit that makes that test actually run (by adding the `Test` prefix).

**Comment by htiennv** [comment]
Thanks @rootulp, I understand why this test fails, because currently we're checking 	
```go
require.Equal(s.T(), context.DeadlineExceeded, err)
```

But in this change I call `GetTx` before call `ctx.Done()`

so it reach error looks like: 
`rpc error: code = DeadlineExceeded desc = context deadline exceeded`

So `require.Equal(s.T(), context.DeadlineExceeded, err)` will fails

I think we can check as follows:
```go
require.Equal(s.T(), strings.Contains(err.Error(), context.DeadlineExceeded.Error()), true)
```
How do you feel about that?

**Comment by rootulp** [comment]
Hmm, slight nit but also appears fixed with:

```go
	assert.Error(t, err)
	assert.Contains(t, err.Error(), context.DeadlineExceeded.Error())
```

File: pkg/user/signer.go [hunk 5]
```diff
@@ -206,36 +206,35 @@ func (s *Signer) BroadcastTx(ctx context.Context, txBytes []byte) (*sdktypes.TxR
 	return resp.TxResponse, nil
 }
 
-// ConfirmTx periodically pings the provided node for the commitment of a transaction by its
-// hash. It will continually loop until the context is cancelled, the tx is found or an error
-// is encountered.
+// ConfirmTx waits for tx to be confirmed on the blockchain.
+// It stops waiting when the context is canceled.
```
Comment (rootulp) [inline pkg/user/signer.go:210]: I think the previous godoc contains information that is still relevant. How about reverting the godoc change?

```go

```
```suggestion
// ConfirmTx periodically pings the provided node for the commitment of a transaction by its
// hash. It will continually loop until the context is cancelled, the tx is found or an error
// is encountered.
```

File: pkg/user/signer.go [hunk 6]
```diff
@@ -206,36 +206,35 @@ func (s *Signer) BroadcastTx(ctx context.Context, txBytes []byte) (*sdktypes.TxR
...
 func (s *Signer) ConfirmTx(ctx context.Context, txHash string) (*sdktypes.TxResponse, error) {
 	txClient := tx.NewServiceClient(s.grpc)
-	timer := time.NewTimer(0)
-	defer timer.Stop()
+
+	s.mtx.RLock()
+	pollTime := s.pollTime
+	s.mtx.RUnlock()
```
Comment (rootulp) [inline pkg/user/signer.go:216]: 👍 seems help to add these read lock/unlocks given 

```go
// SetPollTime sets how often the signer should poll for the confirmation of the transaction
func (s *Signer) SetPollTime(pollTime time.Duration) {
	s.mtx.Lock()
	defer s.mtx.Unlock()
	s.pollTime = pollTime
}
```

File: pkg/user/signer.go [hunk 7]
```diff
@@ -206,36 +206,35 @@ func (s *Signer) BroadcastTx(ctx context.Context, txBytes []byte) (*sdktypes.TxR
...
+
+	pollTicker := time.NewTicker(pollTime)
```
[the remark this answers is not in the document]
Comment (htiennv) [reply pkg/user/signer.go:218]: Oh no, i have poll `GetTx` immediately. 
In fact, the difference is that I didn't check `ctx.Done()` immediately like the previous version.

File: pkg/user/signer.go (hunk 5 above)
Comment (htiennv) [reply pkg/user/signer.go:210]: yes, thank you

File: pkg/user/signer.go (hunk 6 above)
Comment (htiennv) [reply pkg/user/signer.go:216]: I see it was missed in the previous version.

File: pkg/user/signer.go (hunk 5 above)
Comment (htiennv) [reply pkg/user/signer.go]: Done

File: pkg/user/signer_test.go [hunk 8]
```diff
@@ -69,12 +69,15 @@ func (s *SignerTestSuite) TestSubmitTx() {
...
+	t := s.T()
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
 	defer cancel()
-	_, err := s.signer.ConfirmTx(ctx, string("E32BD15CAF57AF15D17B0D63CF4E63A9835DD1CEBB059C335C79586BC3013728"))
-	require.Error(s.T(), err)
-	require.Equal(s.T(), err, context.DeadlineExceeded)
+	_, err := s.signer.ConfirmTx(ctx, "E32BD15CAF57AF15D17B0D63CF4E63A9835DD1CEBB059C335C79586BC3013728")
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), context.DeadlineExceeded.Error())
 }
```
[the remark this answers is not in the document]
Comment (htiennv) [reply pkg/user/signer_test.go:81]: I think it's good. I will add them

**Review by evan-forbes — APPROVED** [review]

File: pkg/user/signer.go (hunk 6 above)
Comment (rootulp) [reply pkg/user/signer.go:217]: thanks for adding! Resolving this.

**Comment by htiennv** [comment]
@rootulp I added some tests for ConfirmTx as you suggest. But I think case 3 (What happens if the tx if found after the second poll) is not necessary and can be ignored.

File: pkg/user/signer_test.go [hunk 9]
```diff
@@ -69,12 +71,44 @@ func (s *SignerTestSuite) TestSubmitTx() {
 	require.EqualValues(t, 0, resp.Code)
 }
 
-func (s *SignerTestSuite) ConfirmTxTimeout() {
-	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+// TestConfirmTx verifies that the ConfirmTx method returns the expected results.
+func (s *SignerTestSuite) TestConfirmTx() {
+	t := s.T()
+
+	// 1. ConfirmTx returns an error when the context time out
```
Comment (rootulp) [inline pkg/user/signer_test.go:78]:
```suggestion
	// 1. ConfirmTx returns an error when the context times out
```

File: pkg/user/signer_test.go [hunk 10]
```diff
@@ -69,12 +71,44 @@ func (s *SignerTestSuite) TestSubmitTx() {
...
+	// 2. ConfirmTx returns an error when tx is not found
+	ctx, cancel = context.WithTimeout(context.Background(), 5*time.Second)
 	defer cancel()
-	_, err := s.signer.ConfirmTx(ctx, string("E32BD15CAF57AF15D17B0D63CF4E63A9835DD1CEBB059C335C79586BC3013728"))
-	require.Error(s.T(), err)
-	require.Equal(s.T(), err, context.DeadlineExceeded)
+	_, err = s.signer.ConfirmTx(ctx, "not found tx")
+	assert.Error(t, err)
+
+	// 3. ConfirmTx has no error if tx is found immediately
```
Comment (rootulp) [inline pkg/user/signer_test.go:91]:
```suggestion
	// 3. ConfirmTx returns no error if tx is found immediately
```

File: pkg/user/signer_test.go [hunk 11]
```diff
@@ -69,12 +71,44 @@ func (s *SignerTestSuite) TestSubmitTx() {
...
+	fee := user.SetFee(1e6)
+	gas := user.SetGasLimit(1e6)
+	msg := bank.NewMsgSend(s.signer.Address(), testnode.RandomAddress().(sdk.AccAddress), sdk.NewCoins(sdk.NewInt64Coin(app.BondDenom, 10)))
+	resp, err := s.submitTxWithoutConfirm([]sdk.Msg{msg}, fee, gas)
+	assert.NoError(t, err)
+	assert.NotNil(t, resp)
+	resp, err = s.signer.ConfirmTx(s.ctx.GoContext(), resp.TxHash)
+	assert.NoError(t, err)
+	assert.Equal(t, resp.Code, uint32(0))
```
Comment (rootulp) [inline pkg/user/signer_test.go:100]: [nit] the order of arguments to assert.Equal is important for debugging. It should be expected then actual so that if this test fails, the logs help debug.

```suggestion
	assert.Equal(t, abci.CodeTypeOK, resp.Code)
```

File: pkg/user/signer_test.go [hunk 12]
```diff
@@ -69,12 +71,44 @@ func (s *SignerTestSuite) TestSubmitTx() {
...
+	// 4. ConfirmTx returns error if the tx is found with a non-zero error code
+	balance := s.queryCurrentBalance(t)
+	// Create a msg send with out of balance, ensure this tx is failed
+	msg = bank.NewMsgSend(s.signer.Address(), testnode.RandomAddress().(sdk.AccAddress), sdk.NewCoins(sdk.NewInt64Coin(app.BondDenom, 1+balance)))
+	resp, err = s.submitTxWithoutConfirm([]sdk.Msg{msg}, fee, gas)
+	assert.NoError(t, err)
+	assert.NotNil(t, resp)
+	resp, err = s.signer.ConfirmTx(s.ctx.GoContext(), resp.TxHash)
+	assert.Error(t, err)
+	assert.NotEqual(t, resp.Code, uint32(0))
```
Comment (rootulp) [inline pkg/user/signer_test.go:111]:
```suggestion
	assert.Equal(t, abci.CodeTypeOK, resp.Code)
```

File: pkg/user/signer_test.go [hunk 13]
```diff
@@ -69,12 +71,44 @@ func (s *SignerTestSuite) TestSubmitTx() {
...
+	resp, err := s.submitTxWithoutConfirm([]sdk.Msg{msg}, fee, gas)
+	assert.NoError(t, err)
+	assert.NotNil(t, resp)
+	resp, err = s.signer.ConfirmTx(s.ctx.GoContext(), resp.TxHash)
+	assert.NoError(t, err)
+	assert.Equal(t, resp.Code, uint32(0))
+
+	// 4. ConfirmTx returns error if the tx is found with a non-zero error code
+	balance := s.queryCurrentBalance(t)
+	// Create a msg send with out of balance, ensure this tx is failed
```
Comment (rootulp) [inline pkg/user/signer_test.go:104]:
```suggestion
	// Create a msg send with out of balance, ensure this tx fails
```

File: pkg/user/signer_test.go [hunk 14]
```diff
@@ -69,12 +71,44 @@ func (s *SignerTestSuite) TestSubmitTx() {
 	require.EqualValues(t, 0, resp.Code)
 }
 
-func (s *SignerTestSuite) ConfirmTxTimeout() {
-	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+// TestConfirmTx verifies that the ConfirmTx method returns the expected results.
```
Comment (rootulp) [inline pkg/user/signer_test.go:74]: In my opinion, "the expected results" is vague. If this comment doesn't provide value, let's remove it entirely.

```suggestion
```

File: pkg/user/signer_test.go [hunk 15]
```diff
@@ -69,12 +71,44 @@ func (s *SignerTestSuite) TestSubmitTx() {
...
+func (s *SignerTestSuite) TestConfirmTx() {
```
Comment (rootulp) [inline pkg/user/signer_test.go:75]: [blocking] Can the test cases in this test be refactored to use table driven tests?

Ref: https://dave.cheney.net/2019/05/07/prefer-table-driven-tests

**Review by rootulp — COMMENTED** [review]
Thanks for adding the test cases! I have a few comments on them

File: pkg/user/signer_test.go (hunk 11 above)
Comment (htiennv) [reply pkg/user/signer_test.go:100]: Thanks

File: pkg/user/signer_test.go (hunk 15 above)
Comment (htiennv) [reply pkg/user/signer_test.go:75]: Thanks for suggest! But I think this seems difficult to do, because each case requires different steps and inputs. For example, some cases need to call submit tx with different msg, some cases do not need to.

What do you think about dividing it into different function: TestConfirmTx_Timeout, TestConfirmTx_NotFoundTx,....?

File: pkg/user/signer_test.go (hunk 12 above)
Comment (htiennv) [reply pkg/user/signer_test.go:111]: `assert.NotEqual(t, abci.CodeTypeOK, resp.Code)`

File: pkg/user/signer_test.go (hunk 15 above)
Comment (rootulp) [reply pkg/user/signer_test.go:74]: Hmm if the setup is distinct for each test case, we could use [this pattern](https://github.com/celestiaorg/celestia-app/blob/ebd733d1db2e75d849a637f68587534447ceccd0/pkg/square/square_test.go#L38-L53) with multiple `t.Run("test case for...")`

Comment (htiennv) [reply pkg/user/signer_test.go:74]: oh thanks, I didn't know about this...

Comment (htiennv) [reply pkg/user/signer_test.go:73]: @rootulp I just push a new commit. Please review it, thank you!

Comment (rootulp) [reply pkg/user/signer_test.go:73]: Awesome, looks great!

**Review by rootulp — APPROVED** [review]
Great work! Thanks for the PR

**Comment by htiennv** [comment]
Thanks!

**Review by evan-forbes — APPROVED** [review]

**Review by staheri14 — APPROVED** [review]
Nice!
